File: frontend/src/app/api/tasks/tasks.ts
import apiClient from '../../../utils/apiClient';

export interface Task {
  id: string;
  title: string;
  description?: string;
  completed: boolean;
  user_id: string;
  created_at: string;
  updated_at: string;
}

export interface TaskCreate {
  title: string;
  description?: string;
  completed?: boolean;
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  completed?: boolean;
}

export const taskApi = {
  // Get all tasks for a user
  getUserTasks: async (userId: string): Promise<{ tasks: Task[]; total: number }> => {
    const response = await apiClient.get(`/${userId}/tasks`);
    return response.data;
  },

  // Create a new task
  createTask: async (userId: string, taskData: TaskCreate): Promise<Task> => {
    const response = await apiClient.post(`/${userId}/tasks`, taskData);
    return response.data;
  },

  // Get a specific task
  getTask: async (userId: string, taskId: string): Promise<Task> => {
    const response = await apiClient.get(`/${userId}/tasks/${taskId}`);
    return response.data;
  },

  // Update a task
  updateTask: async (userId: string, taskId: string, taskData: TaskUpdate): Promise<Task> => {
    const response = await apiClient.put(`/${userId}/tasks/${taskId}`, taskData);
    return response.data;
  },

  // Delete a task
  deleteTask: async (userId: string, taskId: string): Promise<{ message: string }> => {
    const response = await apiClient.delete(`/${userId}/tasks/${taskId}`);
    return response.data;
  },

  // Toggle task completion
  toggleTaskCompletion: async (userId: string, taskId: string): Promise<Task> => {
    const response = await apiClient.patch(`/${userId}/tasks/${taskId}/complete`);
    return response.data;
  },
};

File: backend/src/api/tasks.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlmodel import Session
from ..database import get_session
from ..api.deps import get_current_user
from ..models.user import User
from ..models.task import Task, TaskCreate, TaskUpdate, TaskRead
from ..services.task_service import (
    get_user_tasks,
    create_task_for_user,
    get_task_by_id_and_user,
    update_task_for_user,
    delete_task_for_user,
    toggle_task_completion
)

router = APIRouter()


@router.get("/{user_id}/tasks")
async def list_tasks(
    user_id: str,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_session)
):
    """List all tasks for the authenticated user"""
    if current_user.id != user_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Access denied to other users' tasks"
        )

    tasks = get_user_tasks(session, user_id)
    return {"tasks": tasks, "total": len(tasks)}


@router.post("/{user_id}/tasks", response_model=TaskRead)
async def create_task(
    user_id: str,
    task_create: TaskCreate,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_session)
):
    """Create a new task for the authenticated user"""
    if current_user.id != user_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Access denied to other users' tasks"
        )

    task = create_task_for_user(session, user_id, task_create)
    return task


@router.get("/{user_id}/tasks/{task_id}", response_model=TaskRead)
async def get_task(
    user_id: str,
    task_id: str,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_session)
):
    """Get a specific task for the authenticated user"""
    if current_user.id != user_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Access denied to other users' tasks"
        )

    task = get_task_by_id_and_user(session, task_id, user_id)
    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found"
        )

    return task


@router.put("/{user_id}/tasks/{task_id}", response_model=TaskRead)
async def update_task(
    user_id: str,
    task_id: str,
    task_update: TaskUpdate,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_session)
):
    """Update a specific task for the authenticated user"""
    if current_user.id != user_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Access denied to other users' tasks"
        )

    task = update_task_for_user(session, task_id, user_id, task_update)
    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found"
        )

    return task


@router.delete("/{user_id}/tasks/{task_id}")
async def delete_task(
    user_id: str,
    task_id: str,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_session)
):
    """Delete a specific task for the authenticated user"""
    if current_user.id != user_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Access denied to other users' tasks"
        )

    success = delete_task_for_user(session, task_id, user_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found"
        )

    return {"message": "Task deleted successfully"}


@router.patch("/{user_id}/tasks/{task_id}/complete", response_model=TaskRead)
async def toggle_task_completion_endpoint(
    user_id: str,
    task_id: str,
    current_user: User = Depends(get_current_user),
    session: Session = Depends(get_session)
):
    """Toggle the completion status of a specific task for the authenticated user"""
    if current_user.id != user_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Access denied to other users' tasks"
        )

    task = toggle_task_completion(session, task_id, user_id)
    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found"
        )

    return task